probe/websocket/ws.go

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) 2022, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package websocket
+
+import (
+	"bytes"
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/megaease/easeprobe/global"
+	"github.com/megaease/easeprobe/probe/base"
+)
+
+type WebSocket struct {
+	base.DefaultProbe `yaml:",inline"`
+	URL               string            `yaml:"url" json:"url" jsonschema:"format=uri,title=WebSocket URL,description=WebSocket URL to probe"`
+	Proxy             string            `yaml:"proxy" json:"proxy,omitempty" jsonschema:"format=url,title=Proxy Server,description=proxy to use for the HTTP request"`
+	Headers           map[string]string `yaml:"headers,omitempty" json:"headers,omitempty" jsonschema:"title=HTTP Headers,description=HTTP headers for the initial HTTP request"`
+
+	proxy *url.URL
+}
+
+func (h *WebSocket) Config(gConf global.ProbeSettings) error {
+	kind := "websocket"
+	tag := ""
+	name := h.ProbeName
+	h.DefaultProbe.Config(gConf, kind, tag, name, h.URL, h.DoProbe)
+
+	url, err := url.Parse(h.URL)
+	if err != nil {
+		return err
+	}
+
+	if url.Scheme != "ws" && url.Scheme != "wss" {
+		return fmt.Errorf(`the scheme should be "ws" or "wss", but got: %s`, url.Scheme)
+	}
+
+	if h.Proxy != "" {
+		h.proxy, err = url.Parse(h.Proxy)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+func (h *WebSocket) DoProbe() (bool, string) {
+	wsHeader := make(http.Header)
+	for k, v := range h.Headers {
+		wsHeader.Set(k, v)
+	}
+
+	begin := time.Now()
+	remaining := h.ProbeTimeout
+
+	var dial = websocket.DefaultDialer
+	dial.TLSClientConfig = &tls.Config{
+		InsecureSkipVerify: true,
+	}
+	dial.HandshakeTimeout = remaining
+	if h.proxy != nil {
+		dial.Proxy = func(request *http.Request) (*url.URL, error) {
+			return h.proxy, nil
+		}
+	}
+
+	ws, _, err := dial.Dial(h.URL, wsHeader)
+	if err != nil {
+		return false, err.Error()
+	}
+
+	defer ws.Close()
+
+	pingPongChan := make(chan struct{})
+	ws.SetPongHandler(func(appData string) error {
+		pingPongChan <- struct{}{}
+		return nil
+	})
+
+	// doing nothing else but trigger the read message loop to receive the
+	// Pong and Close messages from the server. exit after ws.Close()
+	wg := sync.WaitGroup{}
```

Comment:
Ideally the wg.Wait is unnecessary. the goroutine will quit shortly.  In case of any extreme condition, the ws connection hangs, doProbe will hang too and not return, to avoid goroutine leak. I was a little knotty when adding the wg. Wait.

Reply:
Removed.